### PR TITLE
[helpers] Provide calls to get free heap and largest available block.

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [x] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [x] BK72xx
- [x] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/helpers.h [hunk 1]
```diff
@@ -710,8 +721,46 @@ template<class T> class RAMAllocator {
...
+        this->flags_ & ALLOC_INTERNAL ? heap_caps_get_free_size(MALLOC_CAP_8BIT | MALLOC_CAP_INTERNAL) : 0;
+    auto max_external =
+        this->flags_ & ALLOC_EXTERNAL ? heap_caps_get_free_size(MALLOC_CAP_8BIT | MALLOC_CAP_SPIRAM) : 0;
+    return max_internal + max_external;
+#elif defined(USE_RP2040)
+    return ::rp2040.getFreeHeap();
+#elif defined(USE_LIBRETINY)
+    return lt_heap_get_free();
+#else
+    return 100000;
```
Comment (guillempages) [inline esphome/core/helpers.h:741]: I'm not sure what will actually cause the else to be compiled; but maybe return some number where it's clear that it's not a right value? E.g. a 0, 1 or 42?

File: esphome/core/helpers.h [hunk 2]
```diff
@@ -684,20 +692,23 @@ template<class T> class RAMAllocator {
...
       ptr = static_cast<T *>(heap_caps_malloc(size, MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT));
     }
-    // Fallback to internal allocation if explicitely requested or no flag is specified
-    if (ptr == nullptr && ((this->flags_ & Flags::ALLOC_INTERNAL) || (this->flags_ & Flags::ALLOC_EXTERNAL) == 0)) {
-      ptr = static_cast<T *>(malloc(size));  // NOLINT(cppcoreguidelines-owning-memory,cppcoreguidelines-no-malloc)
+    if (ptr == nullptr && this->flags_ & Flags::ALLOC_INTERNAL) {
+      ptr = static_cast<T *>(heap_caps_malloc(size, MALLOC_CAP_INTERNAL | MALLOC_CAP_8BIT));
     }
 #else
     // Ignore ALLOC_EXTERNAL/ALLOC_INTERNAL flags if external allocation is not supported
```
Comment (guillempages) [inline esphome/core/helpers.h:714]: Does the comment actually fit the code? The #ifdef is about ESP32, and not really about external allocation, is it?

**Review by guillempages — COMMENTED** [review]
Wasn't there already another PR implementing the INTERNAL/EXTERNAL Flag meaning changes a while ago?

File: esphome/core/helpers.h (hunk 1 above)
Comment (clydebarrow) [reply esphome/core/helpers.h:741]: Host platform. No way to get an actual value, it's just something large. MAX_INT is probably too large.

File: esphome/core/helpers.h (hunk 2 above)
Comment (clydebarrow) [reply esphome/core/helpers.h:714]: ESP32 is the only one that supports external allocation, so it is, yes. There is no #define for external allocation as such.

**Comment by clydebarrow** [comment]
> Wasn't there already another PR implementing the INTERNAL/EXTERNAL Flag meaning changes a while ago?

yes. This doesn't change the meaning, just adds the query calls. But the logic was a bit convoluted before - both flags set meant the same as no flags set, so that correspondence is moved to the constructor rather than having to have an obscure two-part test in multiple places.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**